### PR TITLE
Async local context manager

### PR DESCRIPTION
**Pull request by bizob2828**
## Proposed Release Notes 
 * Added a new context manager that leverages AsyncLocalStorage for async context propagation.
    * This will be available via a feature flag  `config.feature_flag.async_local_context`
    * Alternatively you can set the environment variable of `NEW_RELIC_FEATURE_FLAG_ASYNC_LOCAL_CONTEXT=1`
    * By enabling this feature flag it should make the agent use less memory and CPU.
 
## Links
Closes NEWRELIC-4419

## Details
This new context manager does affect some very rare edge cases.  You can see examples starting here in the [native promises](https://github.com/bizob2828/node-newrelic/blob/async-local-context-manager/test/integration/core/native-promises/native-promises.js#L219) integration tests.  

This does also add 6 additional CI jobs to test unit, integration and versioned tests on async local storage context manager. I'm not sure how I feel about this

### CODE REVIEW

File: third_party_manifest.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "lastUpdated": "Thu Oct 06 2022 11:05:55 GMT-0400 (Eastern Daylight Time)",
+  "lastUpdated": "Wed Oct 12 2022 15:16:03 GMT-0400 (Eastern Daylight Time)",
```
Comment (jordigh) [inline third_party_manifest.json:2]: Why did the timestamp change but nothing else did?

File: test/versioned/mysql2/promises.tap.js [hunk 2]
```diff
@@ -101,21 +101,19 @@ tap.test('mysql2 promises', { timeout: 30000 }, (t) => {
   t.test('query with options object rather than sql', (t) => {
     return helper
       .runInTransaction(agent, () => {
-        return client.query({ sql: 'SELECT 1' })
-      })
-      .then(() => {
-        agent.getTransaction().end()
+        return client.query({ sql: 'SELECT 1' }).then(() => {
+          agent.getTransaction().end()
+        })
```
Comment (jordigh) [inline test/versioned/mysql2/promises.tap.js:106]: These tests were weird, I still don't understand how they ever worked, but the new versions make a lot more sense.

Could now omit a call to `agent.getTransaction()` and just pass it in as the first arg to the callback for `runInTransaction`, slightly cleaner.

File: test/versioned/connect/route.tap.js [hunk 3]
```diff
@@ -28,7 +28,7 @@ test('transaction tests', function (t) {
     t.plan(10)
     const connect = require('connect')
     const { version: pkgVersion } = require('connect/package')
-    let server
+    let server = null
```
Comment (jordigh) [inline test/versioned/connect/route.tap.js:31]: Is this change cosmetic? The variable never seems to be used or checked before it's assigned a non-null value below.

Comment (bizob2828) [reply test/versioned/connect/route.tap.js:31]: good question, most of this was from michael. I can dig

Comment (bizob2828) [reply test/versioned/connect/route.tap.js:31]: it's prob a habit he gets in. no harm in this, I'm going to keep

File: test/versioned/mysql2/promises.tap.js (hunk 2 above)
Comment (bizob2828) [reply test/versioned/mysql2/promises.tap.js:106]: We tend to do the agent.getTransaction() to ensure the current transaction is still the one we're running in. but for this case it's just ending the transaction. I'm going to update but also add an assertion that the transaction we're running in is still the right one

File: third_party_manifest.json (hunk 1 above)
Comment (bizob2828) [reply third_party_manifest.json:2]: so when package.json changes the hooks for the third party update. since no deps changed the timestamp is the only thing that gets updated

File: lib/context-manager/async-local-context-manager.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    this._asyncLocalStorage = new AsyncLocalStorage()
+  }
+
+  /**
+   * Get the currently active context.
+   *
+   * @returns {object} The current active context.
+   */
+  getContext() {
+    return this._asyncLocalStorage.getStore() || EMPTY_CONTEXT
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/context-manager/async-local-context-manager.js:37]: yea I don't see any value in this constant

[the remark this answers is not in the document]
Comment (mrickard) [reply lib/context-manager/async-local-context-manager.js]: Lost contexts get returned as `null` anyway, yeah.

[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/context-manager/async-local-context-manager.js]: It actually returns undefined if not in context

**Review by jordigh — APPROVED** [review]
